### ISSUE TITLE
Bcachefs（system） hang  after making  100+ snapshots 

### ISSUE THREAD

**Issue by bhzhu203** [koverstreet/bcachefs#514]
version f75ef3c588c072ab6444945d8505d0ab90b54e75

````
[ 1195.719246] BUG: kernel NULL pointer dereference, address: 0000000000000000
[ 1195.719640] #PF: supervisor read access in kernel mode
[ 1195.719887] #PF: error_code(0x0000) - not-present page
[ 1195.720129] PGD 0 P4D 0 
[ 1195.720257] Oops: 0000 [#1] PREEMPT SMP PTI
[ 1195.720488] CPU: 0 PID: 2965 Comm: bcachefs Kdump: loaded Tainted: G           O       6.1.0-uksm+ #50 346e44c1ce9f02bb71dda964ff5209e1818a5f27
[ 1195.721066] Hardware name: Alibaba Cloud Alibaba Cloud ECS, BIOS 449e491 04/01/2014
[ 1195.721443] RIP: 0010:bch2_btree_iter_peek_upto+0x50d/0x810 [bcachefs]
[ 1195.721789] Code: f7 d1 48 f7 e1 48 89 c8 48 c1 ea 07 48 89 d6 48 69 d2 cc 00 00 00 48 c1 e6 0c 48 29 d0 48 8d 04 80 48 8d 34 86 e8 83 71 f6 e0 <8b> 00 85 c0 74 05 44 39 e8 72 bc 4c 8b 95 78 ff ff ff 4c 8b 9d 70
[ 1195.722649] RSP: 0018:ffffc900026672c0 EFLAGS: 00010246
[ 1195.722901] RAX: 0000000000000000 RBX: ffffc900026674b8 RCX: 0000000000000000
[ 1195.723225] RDX: 0000000000000000 RSI: 0000000000001000 RDI: 0000000000000000
[ 1195.723581] RBP: ffffc90002667358 R08: ffff8881117b3000 R09: 000000000000000c
[ 1195.723910] R10: 0000000040000382 R11: 0000000000000000 R12: ffffc900026674f8
[ 1195.724235] R13: 00000000ffffff35 R14: ffff88810ab00928 R15: 0000000000000000
[ 1195.724593] FS:  00007f08cbc4ac00(0000) GS:ffff888236e00000(0000) knlGS:0000000000000000
[ 1195.724966] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 1195.725233] CR2: 0000000000000000 CR3: 000000010da66002 CR4: 00000000003706f0
[ 1195.725589] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 1195.725919] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[ 1195.726245] Call Trace:
[ 1195.726368]  <TASK>
[ 1195.726509]  bch2_btree_iter_peek_slot+0x1d7/0x760 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.727000]  ? btree_key_cache_fill+0x70/0x520 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.727437]  btree_key_cache_fill+0x7c/0x520 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.727901]  ? lock_acquire+0x24d/0x2b0
[ 1195.728092]  ? btree_key_cache_fill+0x70/0x520 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.728550]  bch2_btree_path_traverse_cached_slowpath+0x1df/0x1580 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.729073]  bch2_btree_path_traverse_cached+0x1c0/0x240 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.729564]  bch2_btree_path_traverse_one+0x5cc/0xd20 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.730039]  ? bch2_create_trans+0x381/0x730 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.730480]  ? __bch2_btree_path_set_pos+0x1da/0x590 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.730972]  bch2_btree_iter_traverse+0x13b/0x1b0 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.731417]  bch2_create_trans+0x381/0x730 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.731885]  ? bch2_inode_peek+0x8b/0x100 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.732319]  ? bch2_inode_peek+0x8b/0x100 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.732777]  __bch2_create+0x246/0x620 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.733198]  ? __bch2_create+0x175/0x620 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.733631]  bch2_fs_file_ioctl+0x7c0/0xb10 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.734081]  ? bch2_fs_file_ioctl+0x7c0/0xb10 [bcachefs 55e5ef7a52cb8aefa1667a73669e1ee788a33f99]
[ 1195.734514]  ? __fget_files+0xd3/0x170
[ 1195.734728]  __x64_sys_ioctl+0x91/0xc0
[ 1195.734917]  ? __x64_sys_ioctl+0x91/0xc0
[ 1195.735433]  do_syscall_64+0x34/0x80
[ 1195.735922]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
[ 1195.736422] RIP: 0033:0x7f08cbb1aaff
[ 1195.736937] Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <41> 89 c0 3d 00 f0 ff ff 77 1f 48 8b 44 24 18 64 48 2b 04 25 28 00
[ 1195.738356] RSP: 002b:00007ffdd8a85540 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
[ 1195.739037] RAX: ffffffffffffffda RBX: 00007ffdd8a86744 RCX: 00007f08cbb1aaff
[ 1195.739634] RDX: 00007ffdd8a855c0 RSI: 000000004020bc10 RDI: 0000000000000003
[ 1195.740264] RBP: 00007ffdd8a8674a R08: 00005626f99fdf50 R09: 000000007fffffff
[ 1195.740901] R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000001
[ 1195.741503] R13: 0000000000000003 R14: 00005626f7c2a488 R15: 00005626f7d2e900
[ 1195.742136]  </TASK>
[ 1195.742504] Modules linked in: netconsole bcachefs(O) mean_and_variance tcp_diag inet_diag xt_conntrack nft_chain_nat xt_MASQUERADE nf_nat nf_conntrack_netlink nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 xfrm_user xfrm_algo xt_addrtype nft_compat nf_tables nfnetlink br_netfilter bridge stp llc zram binfmt_misc nls_utf8 intel_rapl_msr nls_cp437 intel_rapl_common joydev virtio_console virtio_balloon serio_raw evdev squashfs loop dm_multipath dm_mod ipmi_devintf ipmi_msghandler msr fuse efi_pstore configfs ip_tables x_tables autofs4 raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx raid1 raid0 multipath linear md_mod virtio_net net_failover failover virtio_blk crct10dif_pclmul crc32_pclmul cirrus drm_shmem_helper ghash_clmulni_intel drm_kms_helper sha512_ssse3 sha512_generic drm psmouse virtio_pci i2c_piix4 virtio virtio_pci_legacy_dev aesni_intel crypto_simd cryptd virtio_pci_modern_dev virtio_ring i2c_core floppy button
[ 1195.748462] CR2: 0000000000000000
````

**Comment by bhzhu203** [comment]
```
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_iter_peek_upto+0x50d/0x810
bch2_btree_iter_peek_upto+0x50d/0x810:
bch2_snapshot_parent 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/subvolume.h:27
(已内连入)bch2_snapshot_is_ancestor 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/subvolume.h:66
(已内连入)bch2_btree_iter_peek_upto 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.c:2141
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_iter_peek_slot+0x1d7/0x760
bch2_btree_iter_peek_slot+0x1d7/0x760:
PTR_ERR_OR_ZERO 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/./include/linux/err.h:59
(已内连入)bkey_err 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.h:556
(已内连入)bch2_btree_iter_peek_slot 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.c:2515
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko btree_key_cache_fill+0x70/0x520
btree_key_cache_fill+0x70/0x520:
btree_key_cache_fill 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_key_cache.c:393
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_path_traverse_cached_slowpath+0x1df/0x1580
bch2_btree_path_traverse_cached_slowpath+0x1df/0x1580:
bch2_btree_path_traverse_cached_slowpath 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_key_cache.c:533
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_path_traverse_cached+0x1c0/0x240
bch2_btree_path_traverse_cached+0x1c0/0x240:
bch2_btree_path_traverse_cached 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_key_cache.c:602
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_path_traverse_one+0x5cc/0xd20
bch2_btree_path_traverse_one+0x5cc/0xd20:
bch2_btree_path_traverse_one 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.c:1138
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_create_trans+0x381/0x730
bch2_create_trans+0x381/0x730:
bch2_create_trans 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/fs-common.c:128
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko __bch2_btree_path_set_pos+0x1da/0x590
__bch2_btree_path_set_pos+0x1da/0x590:
__bch2_btree_path_set_pos 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.c:1277
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_btree_iter_traverse+0x13b/0x1b0
bch2_btree_iter_traverse+0x13b/0x1b0:
bch2_btree_iter_traverse 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/btree_iter.c:1681
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_create_trans+0x381/0x730
bch2_create_trans+0x381/0x730:
bch2_create_trans 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/fs-common.c:128
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_inode_peek+0x8b/0x100
bch2_inode_peek+0x8b/0x100:
bch2_inode_peek 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/inode.c:335
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko __bch2_create+0x246/0x620
__bch2_create+0x246/0x620:
__bch2_create 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/fs.c:252 (discriminator 4)
❰bhzhu❙~/source/bcachfs-github/bcachefs-1(git:master)❱✔≻ scripts/faddr2line  fs/bcachefs/bcachefs.ko bch2_fs_file_ioctl+0x7c0/0xb10
bch2_fs_file_ioctl+0x7c0/0xb10:
PTR_ERR_OR_ZERO 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/./include/linux/err.h:59
(已内连入)bch2_ioctl_subvolume_create 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/fs-ioctl.c:415
(已内连入)bch2_fs_file_ioctl 于 /home/bhzhu/source/bcachfs-github/bcachefs-1/fs/bcachefs/fs-ioctl.c:516
````

**Comment by koverstreet** [comment]
Fixed by b95571099b
